roles/openshift_health_checker/openshift_checks/efk_stack.py

```diff
@@ -0,0 +1,604 @@
+        masters = get_var(task_vars, "groups", "masters", default=None) or [None]
+        is_active = logging_deployed and masters[0] == hostname
+        return super(EFKStack, cls).is_active(task_vars) and is_active
+
+    def run(self, tmp, task_vars):
+        '''Check various things and gather errors. Returns: result as hash'''
+        self.logging_namespace = get_var(task_vars, "openshift_logging_namespace", default="logging")
+        pods_for, error = self._get_pods(task_vars)
+        if error:
+            return {"failed": True, "changed": False, "msg": error}
+        check_error = dict(
+            elasticsearch=self.check_elasticsearch(pods_for['es'], task_vars),
+            fluentd=self.check_fluentd(pods_for['fluentd'], task_vars),
+            curator=self.check_curator(pods_for['curator']),
+            kibana=self.check_kibana(pods_for['kibana']),
+        )
+        if not check_error['kibana']:
+            check_error['kibana_route'] = self._check_kibana_route(task_vars)
+
+        errors = [err for err in check_error.values() if err]
+        if errors:
+            errors.insert(0, "The following logging stack issues were found:")
+            return {"failed": True, "changed": False, "msg": "\n-------\n".join(errors)}
+
+        # TODO(lmeyer): run it all again for the ops cluster
+        return {"failed": False, "changed": False, "msg": 'No problems found with logging stack.'}
+
+    def _get_pods(self, task_vars):
+        '''Get all pods and sort them by EFK component. Returns: pods by component, error string'''
+        pod_output = self._exec_oc("get pods -o json", [], task_vars)
+        try:
+            pods = json.loads(pod_output)
+            if not pods or not pods.get('items'):
+                raise ValueError()
+        except ValueError:
+            # successful run but non-parsing data generally means there were no pods in the namespace
+            return None, 'There are no pods in the {} namespace. Is logging deployed?'.format(self.logging_namespace)
+
+        logging_components = ['es', 'fluentd', 'kibana', 'curator']
+        pods_for = {comp: [] for comp in logging_components}
+
+        for pod in pods['items']:
+            component = pod['metadata']['labels'].get('component')
+            if component in logging_components:
+                pods_for[component].append(pod)
+
+        return pods_for, None
+
+    @staticmethod
+    def _not_running_pods(pods):
+        '''Returns: list of pods not in a ready and running state'''
+        return [
+            pod for pod in pods
+            if any(
+                container['ready'] is False
+                for container in pod['status']['containerStatuses']
+            ) or not any(
+                condition['type'] == 'Ready' and condition['status'] == 'True'
+                for condition in pod['status']['conditions']
+            )
+        ]
+
+    def _running_es_pods(self, es_pods):
```

Comment:
It really is just because I want the messaging to be different for each component. Otherwise we could have a single `_errors_from_non_running_pods` method to run for all of them.

I wanted the top-level `check_elasticsearch` to be really easy to read.

I struggled for a long time about where to put messaging logic (what to say about what is observed) versus the evaluating logic it's based on. I did not want every leaf method to generate user-directed errors and have them cascade up, as then the messaging logic is really scattered around and it clutters up leaf methods and makes them hard to reuse. On the other hand I didn't want on huge messaging logic method at the top and didn't want to develop a system of packaging up results just so they can be evaluated legibly several levels up. So I struggled with the level of coupling between evaluating and messaging, and I don't know that I got the balance right, but it seemed worse other ways.